### ISSUE TITLE
Add integration with `serialize` package

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#14731]
### What is the problem this feature will solve?

The [serialize](https://github.com/hgrecco/serialize) package provides a nice unified interface for [json](http://docs.python.org/3/library/json.html), [yaml](http://pyyaml.org/), [shelve](http://docs.python.org/3.6/library/shelve.html), [hdf5](http://www.h5py.org/) and and [dill](https://pypi.python.org/pypi/dill). 

We at Astropy have rolled our own, with incomplete support, for yaml, and not much / any for json. If we support serialize, we get all the formats.

### Describe the desired outcome

- Adopt `serialize` package as an optional dependency, mandatory if you want I/O with JSON, yaml, shelve, and dill (we have our own HDF5)
- Refactor our YAML serialization to work with `serialize` and make our stuff just registering into `serialize`.

**Comment by pllim** [comment]
The package does not seem very active. Last release was over a year ago. There are only 2 contributors. Are you sure this isn't too risky?

**Comment by saimn** [comment]
Serialization for what ? YAML is used for a few specific cases where I don't think it can be replaced by any random (binary) serialization format.

**Comment by mhvk** [comment]
JSON has definitely been requested before - #4604, #5165 - and our YAML interface is arguably somewhat clunky. But I share @pllim's worry about relying on a package with few maintainers that perhaps is not all that widely used...

**Comment by nstarman** [comment]
We could talk to the maintainer (who also maintains `pint`)... see what the future of the package is.
How open they are to help maintaining, maybe?
It would be wonderful for Astropy objects to have a much larger set of I/O options, particularly JSON, and I'm not convinced we should be rolling our own solution. `serialize` is pure python, has no optional dependencies, and looks well designed. Were we to roll our own solution, it would probably look pretty similar to `serialize`.

**Comment by nstarman** [comment]
@hgrecco, what do you think of collaborating on `serialize`?

**Comment by hgrecco** [comment]
@nstarman I am fan of astropy, so I would be happy to help! I always thought it was a pity that we took different paths about units. 

Serialize has not seen too much movement in the last years because it does what it was supposed to do. It is really a thin layer above other serializations packages. Now that typing ecosystem has stabilized, I think it would be good to add typing information.

**Comment by nstarman** [comment]
I agree about units. It would be great to discuss possibilities to create greater interoperability, in a separate discussion 😄.

I'm very on board to add typing.
It would be great to start a working group, perhaps on the Astropy slack(?) to discuss the package.
Are you on the slack?

**Comment by nstarman** [comment]
I sent an invite.

**Comment by hgrecco** [comment]
@mwcraig I was willing to make the requested changes in serialize but need further feedback. We are in the serialize channel on slack

**Comment by MridulS** [comment]
I actually looked a bit into this couple of weeks back and I think one major change required would be to be able to add a hook to `register_class` which lets users create a custom dumpers and loaders. Maybe that means it's a "new astropy format" instead of a new `register_class` but from what I could gather from the `serialize` github repo this seems to be the way to make it backwards compatible? (I'm probably missing some things here).

**Comment by nstarman** [comment]
@mwcraig we could pretty easily add ad-hoc support for astropy types to `serialize`. The issue isn't doing this at an individual level, but migrating all of Astropy's I/O, particularly in a backwards-compatible way.
